### PR TITLE
Use libblockdev for (un)mounting

### PR DESCRIPTION
**Pull request by vojtechtrefny**
Use libblockdev FS plugin `bd_fs_mount` and `bd_fs_unmount` functions instead of `mount` and `umount` programs.

### CODE REVIEW

File: src/udiskslinuxfilesystem.c [hunk 1]
```diff
@@ -1302,23 +1295,22 @@ handle_mount (UDisksFilesystem      *filesystem,
...
-                                                  object,
-                                                  "filesystem-mount", caller_uid,
-                                                  NULL,  /* GCancellable */
-                                                  mount_fstab_as_root ? 0 : caller_uid, /* uid_t run_as_uid */
-                                                  mount_fstab_as_root ? 0 : caller_uid, /* uid_t run_as_euid */
-                                                  &status,
-                                                  &error_message,
-                                                  NULL,  /* input_string */
-                                                  "mount %s",
-                                                  escaped_mount_point_to_use))
```
Comment (vpodzime) [inline src/udiskslinuxfilesystem.c]: Shouldn't we create a simple job here?

File: src/udiskslinuxfilesystem.c [hunk 2]
```diff
@@ -1302,23 +1295,22 @@ handle_mount (UDisksFilesystem      *filesystem,
...
+      if (!mount_fstab_as_root)
         {
-          /* mount(8) exits with status 1 on "incorrect invocation or permissions" - if this is
-           * is so, try as as root */
-          if (!mount_fstab_as_root && WIFEXITED (status) && WEXITSTATUS (status) == 1)
+          BDExtraArg uid_arg = {g_strdup ("run_as_uid"), g_strdup_printf("%d", caller_uid)};
+          BDExtraArg gid_arg = {g_strdup ("run_as_gid"), g_strdup_printf("%d", find_primary_gid (caller_uid))};
+          const BDExtraArg *extra_args[3] = {&uid_arg, &gid_arg, NULL};
+          success = bd_fs_mount (NULL, mount_point_to_use, NULL, NULL, extra_args, &error);
```
Comment (vpodzime) [inline src/udiskslinuxfilesystem.c]: strings from ``uid_arg`` and ``gid_arg`` have to be freed.

File: src/udiskslinuxfilesystem.c [hunk 3]
```diff
@@ -1475,37 +1467,19 @@ handle_mount (UDisksFilesystem      *filesystem,
...
-                                              0,    /* uid_t run_as_euid */
-                                              NULL, /* gint *out_status */
-                                              &error_message,
-                                              NULL,  /* input_string */
-                                              "mount -t %s -o %s %s %s",
-                                              escaped_fs_type_to_use,
-                                              escaped_mount_options_to_use,
-                                              escaped_device,
-                                              escaped_mount_point_to_use))
+  if (!bd_fs_mount (udisks_block_get_device (block), mount_point_to_use, fs_type_to_use, mount_options_to_use, NULL, &error))
```
Comment (vpodzime) [inline src/udiskslinuxfilesystem.c]: Please define a variable for the device name and store the result from ``udisks_block_get_device()`` there.

File: src/udiskslinuxfilesystem.c [hunk 4]
```diff
@@ -1645,27 +1601,20 @@ handle_unmount (UDisksFilesystem      *filesystem,
...
-          /* umount(8) does not (yet) have a specific exits status for
-           * "insufficient permissions" so just try again as root
-           *
-           * TODO: file bug asking for such an exit status
-           */
-          if (!unmount_fstab_as_root && WIFEXITED (status) && WEXITSTATUS (status) != 0)
+          BDExtraArg uid_arg = {g_strdup ("run_as_uid"), g_strdup_printf("%d", caller_uid)};
+          BDExtraArg gid_arg = {g_strdup ("run_as_gid"), g_strdup_printf("%d", find_primary_gid (caller_uid))};
+          const BDExtraArg *extra_args[3] = {&uid_arg, &gid_arg, NULL};
+          success = bd_fs_unmount (mount_point, FALSE, opt_force, extra_args, &error);
```
Comment (vpodzime) [inline src/udiskslinuxfilesystem.c]: Same note here about freeing the strings.

File: src/udiskslinuxfilesystem.c [hunk 5]
```diff
@@ -1690,10 +1639,11 @@ handle_unmount (UDisksFilesystem      *filesystem,
 
           g_dbus_method_invocation_return_error (invocation,
                                                  UDISKS_ERROR,
-                                                 get_error_code_for_umount (status, error_message),
+                                                 get_error_code_for_umount (error->code, error->message),
```
Comment (vpodzime) [inline src/udiskslinuxfilesystem.c]: Is ``error->code`` here the same as what ``status`` used to be?

File: src/tests/dbus-tests/test_80_filesystem.py [hunk 6]
```diff
@@ -238,12 +245,270 @@ class VFATTestCase(UdisksFSTestCase):
...
+        if ret == 0:
+            pipe.send([False, 'Unmount DBus call didn\'t fail but %s seems to be still mounted.' % device])
+            pipe.close()
+            return
+        else:
+            pipe.send([False, '%s was unmounted for UID %d without proper record in fstab' % (device, uid)])
+            pipe.close()
+            return
+
+    def test_mount_fstab_user(self):
```
Comment (vpodzime) [inline src/tests/dbus-tests/test_80_filesystem.py]: Please mark these tests to only run in jenkins.

File: src/tests/dbus-tests/test_80_filesystem.py [hunk 7]
```diff
@@ -238,12 +245,270 @@ class VFATTestCase(UdisksFSTestCase):
...
+        proc = Process(target=self._mount_as_user_fstab, args=(child_conn, int(uid), int(gid), self.vdevs[0]))
+        proc.start()
+        res = parent_conn.recv()
+        parent_conn.close()
+        proc.join()
+
+        if not res[0]:
+            self.fail(res[1])
+
+    def test_mount_fstab_user_fail(self):
```
Comment (vpodzime) [inline src/tests/dbus-tests/test_80_filesystem.py]: Same here.

File: src/udiskslinuxfilesystem.c (hunk 1 above)
Comment (vojtechtrefny) [inline src/udiskslinuxfilesystem.c]: Fixed, thanks.

File: src/udiskslinuxfilesystem.c (hunk 5 above)
Comment (vojtechtrefny) [inline src/udiskslinuxfilesystem.c]: It isn't the same, `status` was `errno` from the syscall but `get_error_code_for_umount` never used it, so I thought it doesn't matter. I've changed the function to take only the message.

**Comment by vojtechtrefny** [comment]
Jenkins, test this please.

**Comment by vojtechtrefny** [comment]
Jenkins, test this please.

**Comment by vojtechtrefny** [comment]
Jenkins, test this please.

**Pull request merged**